### PR TITLE
fix: Handle an extra spaces in class names properly

### LINKED ISSUE #11709: HasStyle.addClassNames() doesn't handle multiple whitespaces
### Description of the bug / feature
```
new Div().addClassNames("   foo  bar    baz")
```
fails with
```
java.lang.IllegalArgumentException: Class name cannot be empty
	at com.vaadin.flow.internal.nodefeature.ElementClassList$ClassListView.validate(ElementClassList.java:45)
	at com.vaadin.flow.internal.nodefeature.ElementClassList$ClassListView.validate(ElementClassList.java:31)
	at com.vaadin.flow.internal.nodefeature.NodeList$SetView.add(NodeList.java:113)
	at com.vaadin.flow.internal.nodefeature.NodeList$SetView.add(NodeList.java:86)
	at com.vaadin.flow.component.HasStyle.addClassNames(HasStyle.java:144)
```
### Expected behavior
The div should have three class names added.
### Actual behavior
The code fails with an exception
### Versions:
    - Vaadin / Flow version: 14.6.8
    - Java version: 8

### PR DESCRIPTION
**Pull request by mshabarov**
## Description

Fixes #11709

## Type of change

- [x] Bugfix
- [ ] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

File: flow-server/src/test/java/com/vaadin/flow/component/HasStyleTest.java [hunk 1]
```diff
@@ -178,6 +178,55 @@ public void testRemoveClassNames() {
...
+        component.addClassNames(null, null);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void removeClassNames_removeEmptyClassName_throws() {
+        HasStyleComponent component = new HasStyleComponent();
+        component.addClassNames(" ");
+    }
+
+    @Test(expected = IllegalArgumentException.class)
```
Comment (caalador) [inline flow-server/src/test/java/com/vaadin/flow/component/HasStyleTest.java:224]: Faulty exception expectation as the method explicitly checks Objects.notNull which throws a NPE

File: flow-server/src/test/java/com/vaadin/flow/component/HasStyleTest.java [hunk 2]
```diff
@@ -178,6 +178,55 @@ public void testRemoveClassNames() {
...
+                0, component.getClassNames().size());
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void addClassNames_addEmptyClassName_throws() {
+        HasStyleComponent component = new HasStyleComponent();
+        component.addClassNames(" ");
+    }
+
+    @Test(expected = IllegalArgumentException.class)
```
Comment (caalador) [inline flow-server/src/test/java/com/vaadin/flow/component/HasStyleTest.java:212]: Faulty exception expectation as the method explicitly checks Objects.notNull which throws a NPE

File: flow-server/src/main/java/com/vaadin/flow/component/HasStyle.java [hunk 3]
```diff
@@ -140,7 +141,14 @@ default Style getStyle() {
      */
     default void addClassNames(String... classNames) {
         for (String rawClassName : classNames) {
-            String[] parts = rawClassName.split(" ");
+            Objects.requireNonNull(rawClassName,
+                    "CSS class names cannot include a null element");
```
Comment (caalador) [inline flow-server/src/main/java/com/vaadin/flow/component/HasStyle.java:145]: The tests would expect here a `if(rawClassName == null) throw new IllegalArgumentException`

File: flow-server/src/main/java/com/vaadin/flow/component/HasStyle.java [hunk 4]
```diff
@@ -157,7 +165,14 @@ default void addClassNames(String... classNames) {
      */
     default void removeClassNames(String... classNames) {
         for (String rawClassName : classNames) {
-            String[] parts = rawClassName.split(" ");
+            Objects.requireNonNull(rawClassName,
+                    "CSS class names cannot include a null element");
```
Comment (caalador) [inline flow-server/src/main/java/com/vaadin/flow/component/HasStyle.java:169]: The tests would expect here a `if(rawClassName == null) throw new IllegalArgumentException`

File: flow-server/src/main/java/com/vaadin/flow/component/HasStyle.java [hunk 5]
```diff
@@ -140,7 +141,14 @@ default Style getStyle() {
      */
     default void addClassNames(String... classNames) {
         for (String rawClassName : classNames) {
-            String[] parts = rawClassName.split(" ");
```
Comment (caalador) [inline flow-server/src/main/java/com/vaadin/flow/component/HasStyle.java:143]: What about:
```
if(rawClassName == null) {
 throw new IllegalArgumentException("CSS class names cannot include a null element");
}
String[] parts = rawClassName.trim().replaceAll(" +", " ").split(" ");
```
This wouldn't throw for a empty class name, but any empty class names wouldn't be added either as the array would be empty

**Review by caalador — CHANGES_REQUESTED** [review]
2 added tests are broken.

File: flow-server/src/test/java/com/vaadin/flow/component/HasStyleTest.java (hunk 1 above)
Comment (mshabarov) [reply flow-server/src/test/java/com/vaadin/flow/component/HasStyleTest.java:224]: Thanks for catching this, I think `Objects.notNull` was added at the last moment, and I didn't recheck the tests.

Comment (mshabarov) [reply flow-server/src/test/java/com/vaadin/flow/component/HasStyleTest.java:224]: Done

File: flow-server/src/test/java/com/vaadin/flow/component/HasStyleTest.java (hunk 2 above)
Comment (mshabarov) [reply flow-server/src/test/java/com/vaadin/flow/component/HasStyleTest.java:212]: Done

File: flow-server/src/main/java/com/vaadin/flow/component/HasStyle.java (hunk 3 above)
Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/component/HasStyle.java]: Done

File: flow-server/src/main/java/com/vaadin/flow/component/HasStyle.java (hunk 4 above)
Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/component/HasStyle.java]: Done

File: flow-server/src/main/java/com/vaadin/flow/component/HasStyle.java (hunk 5 above)
Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/component/HasStyle.java:143]: I see the following drawbacks of going this way:
1. replaceAll(" +", " ") doesn't take into account other whitespace characters, like `\t` which can also be present in the input string.
2. If I pass "" or " " into the method, I get `parts` containing one empty string element, and `getClassNames().add(part);` then throws the empty string exception. So I'd prefer to detect the empty string earlier and throw immediately.

Comment (caalador) [reply flow-server/src/main/java/com/vaadin/flow/component/HasStyle.java:143]: Right split at minimum returns the actual string, but I'm not certain it's a good thing to accept totally faulty class names that contain tabs, carriage returns, new lines, vertical tabs or form feed characters.
Even typing the faulty multiple spaces is a bit questionable and feels like supporting typos.

Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/component/HasStyle.java:143]: Ok, sound reasonable, I changed it to use `.split(" +");`

**Review by caalador — APPROVED** [review]

File: flow-server/src/main/java/com/vaadin/flow/component/HasStyle.java [hunk 6]
```diff
@@ -140,7 +140,16 @@ default Style getStyle() {
...
+            if (rawClassName == null) {
+                throw new IllegalArgumentException(
+                        "CSS class names cannot include a null element");
+            }
+            rawClassName = rawClassName.trim();
+            if (rawClassName.isEmpty()) {
+                throw new IllegalArgumentException(
+                        "CSS class names cannot include an empty class name");
+            }
+            String[] parts = rawClassName.split(" +");
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/component/HasStyle.java:152]: @sonatype-lift help

[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/component/HasStyle.java:152]: @sonatype-lift help

[the remark this answers is not in the document]
Comment (caalador) [reply flow-server/src/main/java/com/vaadin/flow/component/HasStyle.java:152]: @sonatype-lift ignore

The issue is about the fact that if you do a `"hello#".split("#")` you get `[hello]` where as `"#hello".split("#")` will return `["", "hello"]`  and the fix I found for this was to use guava which we don't want to add by default here

**Pull request merged**